src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs

```diff
@@ -51,35 +60,83 @@ public void OnProvidersExecuting(PageRouteModelProviderContext context)
                     // Don't register a duplicate entry for this route.
                     continue;
                 }
 
+                PageSelectorModel.PopulateDefaults(routeModel, routeModel.ViewEnginePath, routeTemplate);
+                context.RouteModels.Add(routeModel);
+            }
+        }
+
+        private void AddAreaPageModels(PageRouteModelProviderContext context)
+        {
+            foreach (var item in _project.EnumerateItems(_pagesOptions.AreaRootDirectory))
+            {
+                if (!IsRouteable(item))
+                {
+                    continue;
+                }
+
+                if (!PageDirectiveFeature.TryGetPageDirective(_logger, item, out var routeTemplate))
+                {
+                    // .cshtml pages without @page are not RazorPages.
+                    continue;
+                }
+
+                if (!PageSelectorModel.TryParseAreaPath(_pagesOptions, item.FilePath, _logger, out var areaResult))
+                {
+                    continue;
+                }
+
                 var routeModel = new PageRouteModel(
                     relativePath: item.CombinedPath,
-                    viewEnginePath: item.FilePathWithoutExtension);
-                PageSelectorModel.PopulateDefaults(routeModel, routeTemplate);
+                    viewEnginePath: areaResult.viewEnginePath)
+                {
+                    RouteValues =
+                    {
+                        ["area"] = areaResult.areaName,
+                    },
+                };
+
+                if (IsAlreadyRegistered(context, routeModel))
+                {
+                    // The CompiledPageRouteModelProvider (or another provider) already registered a PageRoute for this path.
+                    // Don't register a duplicate entry for this route.
+                    continue;
+                }
 
+                PageSelectorModel.PopulateDefaults(routeModel, areaResult.pageRoute, routeTemplate);
                 context.RouteModels.Add(routeModel);
             }
         }
 
-        private bool IsAlreadyRegistered(PageRouteModelProviderContext context, RazorProjectItem projectItem)
+        private bool IsAlreadyRegistered(PageRouteModelProviderContext context, PageRouteModel routeModel)
         {
             for (var i = 0; i < context.RouteModels.Count; i++)
             {
-                var routeModel = context.RouteModels[i];
-                if (string.Equals(routeModel.ViewEnginePath, projectItem.FilePathWithoutExtension, StringComparison.OrdinalIgnoreCase) &&
-                    string.Equals(routeModel.RelativePath, projectItem.CombinedPath, StringComparison.OrdinalIgnoreCase))
+                var existingRouteModel = context.RouteModels[i];
+                if (string.Equals(existingRouteModel.ViewEnginePath, routeModel.ViewEnginePath, StringComparison.OrdinalIgnoreCase) &&
```

Comment:
I don't think we really need to check both of these right? Either one of them is a source of truth

Reply:
Yup. I think I was being over cautious when I was writing this.

Reply:
Having the extra check does mean we don't have to do anything special to deal with the case of the two directories overlapping (since they have the same relative path but different view engine paths). Files in different areas could have the same view engine path, so that's not sufficient by itself.